### PR TITLE
[Python] adding MPI interface

### PR DESCRIPTION
**Pull request by philbucher**
WIP adding MPI inteface

to be discussed in the next meeting

Tricky thing is to pass an `MPI_Comm` from Python

references for passing the `MPI_Comm` from `mpi4py`:
- https://stackoverflow.com/a/62449190
- https://gitlab.com/robertodr/pybind11-mpi4py/-/blob/main/src/pb11mpi.cpp

Also important:
Now I added an additional module `CoSimIOMPI` which has the MPI-stuff, i.e.
~~~py
CoSimIOMPI.ConnectMPI(...)
~~~

I also thought of making `MPI` a submodule of `CoSimIO but then we have again some dependencies like in the MappingApp before I created the MPIExtension
i.e. when having 
~~~py
CoSimIO.MPI.ConnectMPI(...)
~~~
then the `MPI` submodule gets loaded when `CoSimIO` is loaded and hence they are tightly coupled

### CODE REVIEW

**Comment by philbucher** [comment]
I did some more digging in the C Python API and I now think that using a getter function will not work (with reasonable effort as I this seems very complicated)
I searches also how others are doing this, Cython seems to be commonly used but also this appears to be far from trivial

Hence for now I would go with only an interface to `mpi4py` and then add more once we need it / know better the requirements.

I added the `mpi4py` interface as an extension that is disabled by default. I have a custom holder for the `mpi4py` - `MPI_Comm` I think this could also be an option for other libraries (but then again I think it will be custom)

Also I think it is worth to add the python-layer (same as Kratos) such that I can easier integrate submodules, I don't like compiling them in one module, as I fear to run into problems when MPI-libs are not available
Bottom line I suggest the following interface:
~~~py
import CoSimIO
from CoSimIO import MPI as CoSimIOMPI # this one is compiled against MPI (is consistent with usage of mpi4py)
from CoSimIO.MPI import mpi4pyInterface # here I have custom wrappers for the MPI_Comm of mpi4py

settings = CoSimIO.Info()
# set the settings ...

from mpi4py import MPI
mpi4py_comm_holder = mpi4pyInterface.mpi4pyCommHolder(MPI.COMM_WORLD)

CoSimIOMPI.ConnectMPI(settings, mpi4py_comm_holder)
~~~

Note: `mpi4pyCommHolder` has a baseclass `MPICommHolder` in `CoSimIO.MPI`, hence it could also be used for other libs in the future

**Comment by philbucher** [comment]
now the interface looks like this:
~~~py
import CoSimIO
settings = CoSimIO.Info()
from CoSimIO.mpi import ConnectMPI

# without mpi4py:
ConnectMPI(settings)

# with mpi4py
from CoSimIO.mpi.mpi4pyInterface import mpi4pyCommHolder
from mpi4py import MPI

ConnectMPI(settings, mpi4pyCommHolder(MPI.COMM_WORLD))
~~~

I like it very much :)
